File: 04/rb/04.rb
#!/usr/bin/env ruby

nums = Range.new *File.read('input').chomp.split(?-)
good = nums.filter{|i| i =~ /(.)\1/ && i.chars == i.chars.sort}

# part 1
p good.size

# part 2
p good.filter{|i| i.chars.chunk{|x|x}.any?{|_,x| x.size == 2}}.size

File: 04/hs/04.hs
#!/usr/bin/env runhaskell

import Data.List

preprocess =
    map (map length . group) .
    filter (all (uncurry (<=)) . (zip <*> tail)) .
    map show

howMany = (length .) . filter . any
part1 = howMany (>=2)
part2 = howMany (==2)

main = do
    input <- readFile "input"
    putStrLn . show $
        let (a,b) = span (/='-') input
            nums = [read a..read $ tail b] :: [Int]
            counts = preprocess nums
        in map ($ counts) [part1, part2]
